### ISSUE TITLE
np.einsum support

### ISSUE THREAD

**Issue by lukasheinrich** [google/jax#37]
support for generic tensor contractions would cover a large class of computations and also provide a foundation for higher order operations. Perhaps jax could then also be added as a `opt_einsum` backend?

**Comment by mattjj** [comment]
Yes, definitely! And you're right that opt_einsum should be very useful here.

Broadly speaking I think we would want to use opt_einsum to translate an `einsum` call into a sequence of calls into `lax.dot_general`, which models the [XLA DotGeneral HLO](https://www.tensorflow.org/xla/operation_semantics#dotgeneral). But to support all of np.einsum's functionality we'd need to handle a few more cases that aren't exactly tensor contractions, like `np.einsum('ii->i', A)` for extracting a matrix diagonal, `np.einsum('ii->', A)` for computing a trace, and `np.einsum('ij->', A)` for a sum over axes (contracting against an implicit all-ones array, you could say).

Let's leave this issue open to track this feature.

**Comment by mattjj** [comment]
Please reopen if you notice any missing cases!

**Comment by lukasheinrich** [comment]
thanks @mattjj 

I tested with some existing code for which I'd like to use jax

```
import jax.numpy as np
import numpy as onp

second = onp.random.uniform(size = (2,3,4))
first = onp.random.uniform(size = (2,1))
result = np.einsum('sa,shb->shab',first,second)
result.shape #(expected shape: (2, 3, 1, 4))
```

which throws a `StopIteration` error, while changing `import jax.numpy as np` to `import numpy as np
` gives the expected shape

**Comment by mattjj** [comment]
Sounds like a bug!

**Comment by mattjj** [comment]
I made a fix in 997c9c5 that I believe handles the issue. I'll merge it in to master after the CI tests pass. The 'tensor product' side of the code, which your example exercises, was (and is) under-tested, in part because I assumed it would be the easiest to get right (d'oh!).

If you've got more good examples like this, it would be awesome to add them them to the einsum test file! I think the only way to be confident we've covered all the edge cases is to keep adding tests. An even better strategy would be to generate random einsum calls, like a special case of the [JAX generated function tests](https://github.com/google/jax/blob/master/tests/generated_fun_test.py), though that would take some time to put together.

**Comment by mattjj** [comment]
Wow, great idea. Thanks!

**Comment by mattjj** [comment]
I started some work in #147, including with test cases based on the dask test cases. Those tests are really great, and uncovered several issues! Thanks, @sjperkins.

**Comment by mattjj** [comment]
I just merged #147, which handles a lot more cases, but there are a couple remaining that fail. They fail in interesting ways, though. In one case the failure manifests nondeterministically and only in Python 3, and based on an assert check I suspect it could even be a bug in opt_einsum. In the other cases we get an abort from XLA, and so while there might also be issues in the einsum code for those cases, I think the fact that they manifest as aborts indicate that there may be XLA bugs lurking there.

The einsum support seems pretty good now, but we've got to patch up those remaining cases! Thoughts welcome.
